### ISSUE TITLE
rand -out /dev/null gives permission denied

### ISSUE THREAD

**Issue by kroeckx** [openssl/openssl#5590]
I think between the 1.1.0 version and master something changed that it's now opened read-write instead of write only, causing an error.

**Comment by mspncp** [comment]
I can confirm this error. It occurred while testing for #5580:

```
msp@msppc:~/src/committer/openssl$ time util/shlib_wrap.sh apps/openssl rand  -out /dev/null  1000000000
rand: Cannot open output file /dev/null, Permission denied
```

**Comment by levitte** [comment]
Ah, it's the change in #3709 that does it, and it has nothing to do with how the file is opened.  If you look closely at [the added check](https://github.com/openssl/openssl/blob/master/apps/opt.c#L707-L709), the first thing that happens is a check if the directory the file resides in is writable.  `/dev` usually isn't (except for root), hence the error message...

**Comment by t-j-h** [comment]
That added check makes no sense to be where it is in my view as it is checking to see if we can create a *new* file in the directory - not if we can open a file that might already exist ...

**Comment by levitte** [comment]
That check is primarly there to check if we're about to try and write to a directory, i.e. if someone did this:

```
mkdir something
openssl {whatever} -out something
```

**Comment by levitte** [comment]
I'm having a closer look at it to see how that can be expressed better.

**Comment by t-j-h** [comment]
I don't think the *check* is needed at all - the operation of opening the file for write will fail - and that is how it should be. We shouldn't be trying to guess in advance what is going to happen with this sort of check. If you wanted to check that the output file wasn't a directory that is very easy to express - but that isn't what the check is doing ... and I also think it doesn't make sense to do that sort of check either.

**Comment by mspncp** [comment]
In principle, I agree with Tim. Just wanted to note that we should also keep the original problem #3404 in mind, wich #3709 tried to fix.

**Comment by t-j-h** [comment]
@InfoHunter can you have a think about this issue and see what better approach can be made for it.

I think this was fixed in the wrong way - in that if we have applications that aren't handling error conditions and reporting appropriately those applications should be fixed - not a pre-check on directory writability for the output path.

**Comment by levitte** [comment]
Well, I don't quite agree.  The desire expressed in #3404 is exactly to check if the output file can be created before starting any action that will eventually be written to that file.  If said file doesn't exist, how does one check that it's writable?

**Comment by mspncp** [comment]
Open it right away, before starting a lengthy computation.

**Comment by levitte** [comment]
And that creates another issue that has already been discussed and corrected, that if the computation fails or is interrupted, you end up with an output file with invalid content.

**Comment by mspncp** [comment]
I see. One would have to implement it as a transaction: open a tmp file and move on success in order not to overwrite the previous version on error.

**Comment by mspncp** [comment]
Ouch! Now I ended up in a circle!  Well there should be a way...

**Comment by levitte** [comment]
The way is to do pre-checks, like we currently do.  But the current checks need an update.

**Comment by levitte** [comment]
Have a look at #5592

**Comment by InfoHunter** [comment]
> @InfoHunter can you have a think about this issue and see what better approach can be made for it.

I tend to update the current check logic as @levitte has already proposed in #5592. The original issue has 2 specific problems to address: 1. don't give a true-run if the final output is impossible (can't write or can't create files...) to avoid flogging a dead horse, and 2. don't really create or truncate file before a true-run to avoid invalid content.

So the solution came up was to add some pre-check when the `-out` option received its value, and it seems it was rational to do so...

**Comment by petrovr** [comment]
Dear all following command ${OPENSSL} crl -out /dev/null -in "${FILE}" -hash works fine in my regression test for about 15 years. For protocol you could see command in diffs for http://roumenpetrov.info/openssh/download-20040819.html#get_f  (2003-01-30 12:14:15+0200)

If seems to me something changed in master last month adds regression.

The commit title " Check directory is able to create files for various -out option" look quite suspicious.
This is my favourite for regression. 

Command "echo 123 > /dev/null" work fine for normal user and does not require write directory permission. 
I change permission of /dev/ to be writable by user user in regression test and as expected "crl" regression test pass.

Please revert: " Check directory is able to create files for various -out option
    
    This is to address issue #3404, only works in Unix-like platforms
    
    Reviewed-by: Matt Caswell <matt@openssl.org>
    Reviewed-by: Richard Levitte <levitte@openssl.org>
    (Merged from https://github.com/openssl/openssl/pull/3709)"

Issue  #3404 does not require any action. I wonder why is not rejected. It is user mistake.
- (1) User is responsible and provide good path for output file. 
- (2) Also is expected program X do not do any output file operation until whole data is ready to be written.
Program X cannot test  file operation in advance -  access to file or directory could be changed between check and output time.

Hmm, may be proper action is to reopen issue  #3404.

**Comment by mspncp** [comment]
@petrovr thanks for your report. There is currently an ongoing discussion on #5592 to fix this issue. Maybe you could try whether that patch works for you?

**Comment by FdaSilvaYY** [comment]
Fixed by PR #5590 with 242fcd695db6225ef98c5ad084e6f15ec5953158.

**Comment by mspncp** [comment]
> Fixed by PR #5590 with 242fcd6.

You meant PR #5737. But yes, I think it's fixed now.
